File: first_non_repeating_element_using_a_hash_map/fnreuahm.ts
function firstNonRepeating<T>(elements: T[]): T | null {
    const counts = new Map<T, number>();

    for (const item of elements) {
        const currentCount = counts.get(item);

        if (currentCount === undefined) {
            counts.set(item, 1);
        } else {
            counts.set(item, currentCount + 1);
        }
    }

    for (const item of elements) {
        if (counts.get(item) === 1) {
            return item;
        }
    }

    return null;
}

const words = ["a", "b", "a", "c", "b"];
const numbers = [4, 5, 4, 6, 5];
const allRepeated = [1, 1, 2, 2, 3, 3];

console.log(firstNonRepeating(words));
console.log(firstNonRepeating(numbers));
console.log(firstNonRepeating(allRepeated));

File: first_non_repeating_element_using_a_hash_map/fnreuahm.py
from collections.abc import Hashable
from typing import TypeVar

T = TypeVar("T", bound=Hashable)


def first_non_repeating(elements: list[T]) -> T | None:
    counts: dict[T, int] = {}

    for item in elements:
        if item in counts:
            counts[item] += 1
        else:
            counts[item] = 1

    for item in elements:
        if counts[item] == 1:
            return item

    return None


words = ["a", "b", "a", "c", "b"]
numbers = [4, 5, 4, 6, 5]
all_repeated = [1, 1, 2, 2, 3, 3]

print(first_non_repeating(words))
print(first_non_repeating(numbers))
print(first_non_repeating(all_repeated))
